Spigot-API-Patches/0243-Added-SoundBuilder.patch

```diff
@@ -0,0 +1,331 @@
++import org.jetbrains.annotations.Nullable;
++
++import java.util.Collection;
++import java.util.List;
++
++/**
++ * Helps prepare a sound to be sent to players.
++ */
++public interface SoundBuilder {
++
++    /**
++     * Plays the sound to all receiving players (or all). This method is safe to use
++     * Asynchronously
++     *
++     * @return a reference to this object.
++     */
++    SoundBuilder play();
++
++    /**
++     * @return The sound going to be sent
++     */
++    @NotNull
++    Sound sound();
++
++    /**
++     * Changes what sound will be sent
++     *
++     * @param sound The sound
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder sound(@NotNull Sound sound);
++
++    /**
++     * @return The category under which the sound will be played
++     */
++    @Nullable
++    SoundCategory category();
++
++    /**
++     * Changes under which category the sound will be played
++     */
++    @NotNull
++    SoundBuilder category(SoundCategory category);
++
++    /**
++     * @return The volume level of the sound played
++     */
++    @NotNull
++    Float volume();
++
++    /**
++     * Changes the volume level of the sound played
++     *
++     * @param volume The volume
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder volume(@NotNull Float volume);
++
++    /**
++     * @return The pitch of the sound played
++     */
++    @NotNull
++    Float pitch();
++
++    /**
++     * Changes the pitch of the sound played
++     *
++     * @param pitch The pitch
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder pitch(@NotNull Float pitch);
++
++    /**
++     * @return List of players who will receive the sound, or null for all in world
++     */
++    @Nullable
++    List<Player> receivers();
++
++    /**
++     * Example use:
++     *
++     * builder.receivers(16)
++     *
++     * @return If this sound is going to be sent to someone
++     */
++    boolean hasReceivers();
++
++    /**
++     * Sends this sound to all players in the world. This is rather silly and you should likely not
++     * be doing this.
++     *
++     * Just be a logical person and use receivers by radius or collection.
++     *
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder allPlayers();
++
++    /**
++     * @param receivers List of players to receive this sound, or null for all players in the
++     * world
++     * @return a reference to this object.
++     * @deprecated The use of list's is not appreciated
++     */
++    @NotNull
++    @Deprecated
++    SoundBuilder receivers(@Nullable List<Player> receivers);
++
++    /**
++     * @param receivers List of players to receive this sound, or null for all players in the
++     * world
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder receivers(@Nullable Collection<Player> receivers);
```

Comment:
Why allow null values? We already have a method that does what a null value does: set the receivers to all players.

Reply:
So that you can more smartly add things for your own plugin without having to do if:null on the plugin side.

Reply:
null should be to clear the collection.